src/lib/concurrency/pipe_lib/swappable_strict_pipe.ml

```diff
@@ -0,0 +1,234 @@
+open Core_kernel
+open Async_kernel
+
+(** short-lived sink: a writer for a pipe which was created upon
+    a call to [swap_reader]. *)
+type 'data_in_pipe short_lived_sink_t =
+  ('data_in_pipe, Strict_pipe.synchronous, unit Deferred.t) Strict_pipe.Writer.t
+
+type ('data_in_pipe, 'write_return) t =
+  | Swappable :
+      { long_lived_writer :
+          ('data_in_pipe, 'pipe_kind, 'write_return) Strict_pipe.Writer.t
+            (** If [termination_signal] is filled, the swappable pipe is
+          set to gracefully terminate.
+        *)
+      ; termination_signal : unit Ivar.t
+            (** mutable variable [next_short_lived_sink] is only written from
+          within the [background_thread]. It contains an [Ivar.t] of a pair.
+          First element of the pair is short lived sink which will become the
+          next sink. Second element of the pair is a "push-back" of type
+          [unit Ivar.t] which is filled once the swapping is done. *)
+      ; mutable next_short_lived_sink :
+          ('data_in_pipe short_lived_sink_t * unit Ivar.t) Ivar.t
+      }
+      -> ('data_in_pipe, 'write_return) t
+
+(** state of the swappable pipe
+
+    Used by the background thread, every iteration of the loop
+    returns an updated state (fields [short_lived_sink] and [data_unconsumed]
+    are updated).
+*)
+type ('data_in_pipe, 'write_return) state_t =
```

Comment:
NOTE for myself: This DS now only owns the downstream short_lived_sink(which is a writer), and it no longer owns the upperstream.